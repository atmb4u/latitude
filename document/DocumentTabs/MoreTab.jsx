/**
 * TEAM: frontend_infra
 * @flow
 */

import * as React from "react";
import {StyleSheet, css} from "aphrodite";
import Dotdotdot from "react-dotdotdot";
import PopupWithClickAway from "../../popup/PopupWithClickAway";
import Text from "../../Text";
import Icon from "../../Icon";
import colors from "../../styles/colors";
import Clickable from "../../base_candidate/Clickable";
import typeof DocumentTab from "./DocumentTab";
import sizes, {type Size} from "./sizes";

type Props = {|
  +children: React.ChildrenArray<React.Element<DocumentTab>>,
  +width: number,
  +selectedLabel: string | null,
  +size: Size,
|};

export const ListPopupContext = React.createContext<{
  togglePopup: () => void,
}>({
  togglePopup: () => {},
});

function MoreTab({width, selectedLabel, size, children}: Props) {
  return (
    <PopupWithClickAway>
      {(Target, Popup, {togglePopup, isOpen}) => (
        <>
          <Target>
            <Clickable onClick={togglePopup}>
              <div
                className={css(
                  styles.container,
                  !!selectedLabel && styles.selected
                )}
                style={{width, height: sizes[size]}}
              >
                <Text
                  color={selectedLabel ? "grey60" : "blue30"}
                  weight="bold"
                  wordBreak="break-word"
                >
                  {selectedLabel ? (
                    <Dotdotdot clamp={size !== "s" ? 2 : 1}>
                      {selectedLabel}
                    </Dotdotdot>
                  ) : (
                    "More"
                  )}
                </Text>
                <div className={css(styles.buttonContainer)}>
                  <Icon iconName="downOpen" color="blue30" />
                </div>
              </div>
            </Clickable>
          </Target>
          {isOpen ? (
            <Popup>
              <div
                className={css(styles.listContainer)}
                style={{width: width + 2}}
              >
                <ListPopupContext.Provider value={{togglePopup}}>
                  {children}
                </ListPopupContext.Provider>
              </div>
            </Popup>
          ) : null}
        </>
      )}
    </PopupWithClickAway>
  );
}

const styles = StyleSheet.create({
  container: {
    display: "flex",
    position: "relative",
    justifyContent: "center",
    alignItems: "center",
    padding: "4px 4px 4px 20px",
    background: colors.white,
  },
  listContainer: {
    marginTop: "4px",
    minWidth: "100%",
    border: `1px solid ${colors.grey30}`,
    background: "white",

    ":nth-child(1n) > :not(:last-child)": {
      borderBottom: `1px solid ${colors.grey30}`,
    },
  },
  buttonContainer: {
    display: "flex",
    justifyContent: "center",
    minWidth: "28px",
  },
  selected: {
    ":after": {
      content: '""',
      position: "absolute",
      left: "0",
      bottom: "0",
      width: "100%",
      height: "2px",
      backgroundColor: colors.blue30,
    },
  },
});

export default MoreTab;
