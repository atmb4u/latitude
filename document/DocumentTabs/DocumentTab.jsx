/**
 * TEAM: frontend_infra
 * @flow
 */

import * as React from "react";
import {StyleSheet, css} from "aphrodite";
import Dotdotdot from "react-dotdotdot";
import PopupWithClickAway from "../../popup/PopupWithClickAway";
import GeneralPopover from "../../popover/GeneralPopover";
import Button from "../../button/Button";
import IconButton from "../../button/IconButton";
import Tooltip from "../../Tooltip";
import Text from "../../Text";
import Clickable from "../../base_candidate/Clickable";
import colors from "../../styles/colors";
import sizes, {type Size} from "./sizes";
import {ListPopupContext} from "./MoreTab";

type Props = {|
  +children: string,
  +isOverflow: boolean,
  +width: number,
  +isSelected: boolean,
  +onClick: () => void,
  +onDelete: () => void,
  +size: Size,
|};

function DocumentTab({
  children,
  isOverflow,
  width,
  size,
  isSelected,
  onClick,
  onDelete,
}: Props) {
  const listPopupContext = React.useContext(ListPopupContext);

  return (
    <PopupWithClickAway>
      {(Target, Popup, {togglePopup}) => (
        <>
          <Target>
            <Clickable
              onClick={() => {
                onClick();
                listPopupContext.togglePopup();
              }}
            >
              {/* $FlowFixMe(dirak): Clickable effectively wraps a div here */}
              <Tooltip
                placement={isOverflow ? "right" : "bottom"}
                mouseEnterDelay={0.5}
                overlay={
                  <div style={{width, wordWrap: "break-word"}}>{children}</div>
                }
              >
                <div
                  className={css(
                    styles.container,
                    !isOverflow && styles.underline,
                    isOverflow && isSelected && styles.greyBackground,
                    isSelected && styles.selected
                  )}
                  style={{width, height: sizes[size]}}
                >
                  <div
                    className={css(styles.textContainer)}
                    // width offset by container margins
                    style={{width: width - 48}}
                  >
                    <Dotdotdot clamp={size !== "s" ? 2 : 1}>
                      {children}
                    </Dotdotdot>
                  </div>

                  <div
                    className={css(
                      styles.buttonContainer,
                      !isSelected && styles.hidden
                    )}
                  >
                    <IconButton
                      iconName="trash"
                      type="button"
                      intent="none"
                      kind="blank"
                      onClick={e => {
                        togglePopup();
                        e.stopPropagation();
                      }}
                    />
                  </div>
                </div>
              </Tooltip>
            </Clickable>
          </Target>
          <Popup placement={isOverflow ? "right-start" : "bottom-start"}>
            <div style={{margin: isOverflow ? "0 8px" : "8px 0"}}>
              <GeneralPopover
                title="Delete Shipping Order"
                buttons={[
                  <Button key={1} label="Cancel" onClick={togglePopup} />,
                  <Button
                    key={2}
                    label="Delete"
                    intent="danger"
                    kind="solid"
                    onClick={() => {
                      onDelete();
                      togglePopup();
                    }}
                  />,
                ]}
              >
                <div className={css(styles.popoverBody)}>
                  <Text display="inline">
                    Are you sure you want to delete <b>{children}</b>?
                  </Text>
                  <br />
                  <Text display="inline">
                    You will not be able to access this file once it is deleted.
                    Data from this file will be deleted as well.
                  </Text>
                  <br />
                </div>
              </GeneralPopover>
            </div>
          </Popup>
        </>
      )}
    </PopupWithClickAway>
  );
}

const styles = StyleSheet.create({
  container: {
    display: "flex",
    position: "relative",
    alignItems: "center",
    padding: "4px 0 4px 20px",
    cursor: "pointer",
    background: colors.white,
    ":hover": {
      background: colors.grey10,
      ":after": {
        backgroundColor: colors.grey30,
      },
    },
    ":hover > :nth-child(2)": {
      display: "flex",
    },

    ":active": {
      background: colors.grey20,
      ":after": {
        backgroundColor: colors.grey40,
      },
    },
  },
  underline: {
    ":after": {
      content: '""',
      position: "absolute",
      left: "0",
      bottom: "0",
      right: "0",
      height: "2px",
      backgroundColor: colors.white,
    },
  },
  selected: {
    fontWeight: "bold",
    ":after": {
      backgroundColor: colors.blue30,
    },
    ":hover": {
      ":after": {
        backgroundColor: colors.blue30,
      },
    },
    ":active": {
      ":after": {
        backgroundColor: colors.blue30,
      },
    },
  },
  textContainer: {
    display: "block",
    wordWrap: "break-word",
  },
  buttonContainer: {
    display: "flex",
    justifyContent: "center",
    minWidth: "28px",
    height: "100%",
    alignSelf: "flex-start",
  },
  hidden: {
    display: "none",
  },
  greyBackground: {
    background: colors.grey20,
  },
  popoverBody: {
    display: "flex",
    flexDirection: "column",
    wordWrap: "break-word",
  },
});

export default DocumentTab;
