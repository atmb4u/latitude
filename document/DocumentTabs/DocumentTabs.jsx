/**
 * TEAM: frontend_infra
 * @flow
 */

import * as React from "react";
import {StyleSheet, css} from "aphrodite";
import Icon from "../../Icon";
import colors from "../../styles/colors";
import DocumentTab from "./DocumentTab";
import MoreTab from "./MoreTab";
import type {Size} from "./sizes";

const MIN_ELEMENT_WIDTH = 130;
const MAX_ELEMENT_WIDTH = 200;
const ADD_BUTTON_WIDTH = 52;

type Props = {|
  /** A list of document names to be displayed. */
  +documentNameOptions: $ReadOnlyArray<{+key: number, +name: string}>,
  /** The key of the currently selected document. */
  +selectedKey: number,
  /** A callback called when the add button is pressed. */
  +onAdd: () => void,
  /** A callback called when a tab is selected. */
  +onSelect: (key: number) => void,
  /** A callback called when a document is deleted. */
  +onDelete: (key: number) => void,
  /** The size of document tabs. */
  +size?: Size,
|};

type Dimensions = {|
  tabWidth: number,
  slices: number,
|};

function calculateWidths(
  containerWidth: number,
  minWidth: number,
  maxWidth: number,
  elementCount: number
): Dimensions {
  if (elementCount * maxWidth < containerWidth) {
    return {tabWidth: maxWidth, slices: elementCount};
  }

  const maxSlicesCount = Math.floor(containerWidth / minWidth);
  const slices = Math.min(maxSlicesCount, elementCount);

  return {
    tabWidth: containerWidth / slices,
    slices,
  };
}

/**
 * @short A tab header for managing multiple documents
 * @brandStatus V2
 * @status Beta
 * @category Documents
 */
function DocumentTabs({
  documentNameOptions,
  selectedKey,
  onAdd,
  onSelect,
  onDelete,
  size = "m",
}: Props) {
  const containerRef = React.useRef<HTMLDivElement | null>(null);
  const [dimensions, setDimensions] = React.useState({
    tabWidth: 0,
    slices: 0,
  });
  const {tabWidth, slices} = dimensions;

  React.useEffect(() => {
    if (containerRef.current !== null) {
      const dimensions = calculateWidths(
        // container size offset by size of addition button
        containerRef.current.offsetWidth - ADD_BUTTON_WIDTH,
        MIN_ELEMENT_WIDTH,
        MAX_ELEMENT_WIDTH,
        documentNameOptions.length
      );

      setDimensions(dimensions);
    }
  }, [documentNameOptions]);

  const isOverflowing = documentNameOptions.length > slices;
  const nonOverflowCount = isOverflowing ? slices - 1 : slices;

  return (
    <div
      className={css(styles.container, tabWidth === 0 && styles.hidden)}
      style={{display: documentNameOptions.length !== 0 ? "flex" : "none"}}
      ref={containerRef}
    >
      {documentNameOptions.slice(0, nonOverflowCount).map(({name, key}) => (
        <DocumentTab
          isOverflow={false}
          width={tabWidth}
          isSelected={key === selectedKey}
          onClick={() => onSelect(key)}
          onDelete={() => onDelete(key)}
          size={size}
          key={key}
        >
          {name}
        </DocumentTab>
      ))}

      {isOverflowing ? (
        <MoreTab
          width={tabWidth}
          size={size}
          selectedLabel={
            selectedKey < nonOverflowCount
              ? null
              : (documentNameOptions.find(({key}) => key === selectedKey) || {})
                  .name
          }
        >
          {documentNameOptions.slice(nonOverflowCount).map(({name, key}) => (
            <DocumentTab
              isOverflow={true}
              width={tabWidth}
              isSelected={key === selectedKey}
              onClick={() => onSelect(key)}
              onDelete={() => onDelete(key)}
              size={size}
              key={key}
            >
              {name}
            </DocumentTab>
          ))}
        </MoreTab>
      ) : (
        undefined
      )}

      {/* eslint-disable react/forbid-elements */}
      <button
        className={css(styles.addDocumentButton)}
        type="button"
        onClick={onAdd}
      >
        <Icon iconName="plus" color="blue30" />
      </button>
    </div>
  );
}

const styles = StyleSheet.create({
  container: {
    flexDirection: "row",
    width: "100%",
    ":nth-child(1n) > *": {
      borderRight: `1px solid ${colors.grey30}`,
    },
  },
  addDocumentButton: {
    display: "flex",
    justifyContent: "center",
    alignItems: "center",
    width: ADD_BUTTON_WIDTH,
    border: "none",
    outline: "none",
    padding: "0",
    margin: "0",
  },
  hidden: {
    visibility: "hidden",
  },
});

export default DocumentTabs;
