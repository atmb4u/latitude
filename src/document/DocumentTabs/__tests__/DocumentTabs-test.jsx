/**
 * TEAM: frontend_infra
 *
 * @flow
 */

import React from "react";
import {shallow} from "enzyme";
import DocumentTabs from "../DocumentTabs";

describe("DocumentTabs", () => {
  it("renders", () => {
    expect(
      shallow(
        <DocumentTabs
          documentNameOptions={["doc1.pdf", "doc2.pdf"].map((name, key) => ({
            name,
            key,
          }))}
          selectedKey={0}
          onSelect={() => {}}
          onAdd={() => {}}
          onDelete={() => {}}
        />
      )
    ).toMatchSnapshot();
  });
});
