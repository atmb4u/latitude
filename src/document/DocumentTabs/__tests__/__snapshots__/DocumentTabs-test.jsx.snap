// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocumentTabs renders 1`] = `
<div
  className="container_zhnr9x-o_O-hidden_gxjeqd"
  style={
    Object {
      "display": "flex",
    }
  }
>
  <DocumentTab
    isOverflow={false}
    isSelected={true}
    key="0"
    onClick={[Function]}
    onDelete={[Function]}
    size="m"
    width={0}
  >
    doc1.pdf
  </DocumentTab>
  <MoreTab
    selectedLabel="doc1.pdf"
    size="m"
    width={0}
  >
    <DocumentTab
      isOverflow={true}
      isSelected={false}
      key="1"
      onClick={[Function]}
      onDelete={[Function]}
      size="m"
      width={0}
    >
      doc2.pdf
    </DocumentTab>
  </MoreTab>
  <button
    className="addDocumentButton_karj7w"
    onClick={[Function]}
    type="button"
  >
    <Icon
      alignment="baseline"
      color="blue30"
      deprecatedAllowColorInheritance={true}
      iconName="plus"
      isARIAAccessible={false}
    />
  </button>
</div>
`;
